mindmeld/text_preparation/text_preparation_pipeline.py

```diff
@@ -0,0 +1,566 @@
+                For Example:
+                    norm_token = {
+                        "entity": "order",
+                        "raw_entity": "order",
+                        "raw_token_index": 1,
+                        "raw_start": 1
+                    }
+        """
+        raw_tokens = self.tokenizer.tokenize(text)
+        normalized_tokens = []
+        for i, raw_token in enumerate(raw_tokens):
+            if not raw_token["text"]:
+                continue
+            normalized_text = self._normalize_text(raw_token["text"])
+            # We tokenize the post-norm text and split the entity if possible
+            # Ex: normalize("o'clock") -> "o clock" -> ["o", "clock"]
+            normalized_texts = [t["text"] for t in self.tokenize(normalized_text)]
+
+            if len(normalized_texts) > 0:
+                for token_text in normalized_texts:
+                    normalized_tokens.append(
+                        {
+                            "entity": token_text,
+                            "raw_entity": raw_token["text"],
+                            "raw_token_index": i,
+                            "raw_start": raw_token["start"],
+                        }
+                    )
+        return normalized_tokens
+
+    def get_normalized_tokens_as_tuples(self, text):
+        """Gets normalized tokens from input text and returns the result as a tuple.
+
+        Args:
+            text (str): Text to normalize.
+        Returns:
+            normalized_tokens_as_tuples (Tuple(str)): A Tuple of normalized tokens.
+        """
+        return tuple(t["entity"] for t in self.tokenize_and_normalize(text))
+
+    def stem_word(self, word):
+        """
+        Gets the stem of a word. For example, the stem of the word 'fishing' is 'fish'.
+
+        Args:
+            words (List[str]): List of words to stem.
+
+        Returns:
+            stemmed_words (List[str]): List of stemmed words.
+        """
+        return self.stemmer.stem_word(word)
+
+    @staticmethod
+    def find_mindmeld_annotation_re_matches(text):
+        """
+        Args:
+            text (str): The string to find mindmeld annotation instances
+                (" {entity_text|entity_type} ")
+        Returns:
+            matches (List[sre.SRE_Match object]): Regex match objects.
+        """
+        return list(MINDMELD_ANNOTATION_PATTERN.finditer(text))
+
+    @staticmethod
+    def modify_around_annotations(text, function, keep_space_before_annotation=False):
+        """Applied a function around the mindmeld annotation.
+
+        function(pre_entity_text) + { + function(entity_text) + |entity_name}
+            + function(post_entity_text)
+
+        Args:
+            text (str): Original sentence with markup to modify.
+            function (function): Function to apply around the annotation
+            keep_space_before_annotation (bool): If True and if there is a space before entity
```

Comment:
Unclear why we have this. The logic should be that if the previous char is a space ie "abc {xyz}", then the space should be preserved, so it's implicit.

Reply:
Preprocessors now process the input text directly